### PR TITLE
Changing Webhook behavior

### PR DESCRIPTION
**Pull request by Andreagit97**
# Description

When a new `Pod` is deployed in a `Namespace` labeled with `liqo.io/scheduling="true"`, the Webhook must set the right `Pod Affinity` according to what the admin has specified in the `NamespaceOffloading.Spec.ClusterSelector` field. If the user deploys a `Pod` with some affinities, these are merged with the ones selected by the administrator in the `ClusterSelector`; if there is a conflict between the two sets of `Affinity` the `Pod` remains pending until someone deletes it.

### CODE REVIEW

File: test/unit/webhook/manage_affinity_test.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
+package webhookLogic
```
Comment (palexster) [inline test/unit/webhook/manage_affinity_test.go:1]: Tests should be formatted with Ginkgo/Gomega libraries.

File: pkg/mutate/manage_nodeaffinity.go [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+// ManageAffinity sets right Pod Affinity according to the PodOffloadingStrategy specified in NamespaceOffloading resource.
+func ManageAffinity(namespaceOffloading *offv1alpha1.NamespaceOffloading, pod *corev1.Pod) {
+	// This newNodeSelector will substitute the old nodeSelector if the strategy is not "LocalPodOffloadingStrategy".
+	newNodeSelector := &corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}}
+
+	// 1 - Fill the newNodeSelector according to the ClusterSelector field of NamespaceOffloading.
+	switch {
+	case namespaceOffloading.Spec.PodOffloadingStrategy == offv1alpha1.LocalAndRemotePodOffloadingStrategyType:
+		// 1 - Liqo toleration
+		klog.Info("LocalAndRemotePodOffloadingStrategy")
```
Comment (palexster) [inline pkg/mutate/manage_nodeaffinity.go:25]: Some suggestions:
1. We should understand which resource is concerned by this log
2. Maybe it should be printed only when the verbosity is increased.

File: pkg/mutate/manage_nodeaffinity.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+		pod.Spec.Tolerations = append(pod.Spec.Tolerations, corev1.Toleration{
+			Key:      liqoconst.VirtualNodeTolerationKey,
+			Operator: corev1.TolerationOpExists,
+			Effect:   corev1.TaintEffectNoExecute,
+		})
+
+		klog.Info("toleration inserted")
```
Comment (palexster) [inline pkg/mutate/manage_nodeaffinity.go:33]: Ditto.

File: pkg/mutate/manage_nodeaffinity.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+// todo: if the clusterSelcetor is empty at NamespaceOffloading creation time, in the NamespaceOffloading Controller
+//       we have to set with a default matchexpression for virtual nodes only, so this is never empty
+
+// This function is global in order to access it with test in test/unit/webhookLogic
+
+// ManageAffinity sets right Pod Affinity according to the PodOffloadingStrategy specified in NamespaceOffloading resource.
+func ManageAffinity(namespaceOffloading *offv1alpha1.NamespaceOffloading, pod *corev1.Pod) {
+	// This newNodeSelector will substitute the old nodeSelector if the strategy is not "LocalPodOffloadingStrategy".
+	newNodeSelector := &corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}}
```
Comment (palexster) [inline pkg/mutate/manage_nodeaffinity.go:19]: I think this may be acceptable in the first place but then I would tend to merge the existing selector with the Liqo one. @giorio94 what do you think about it?

Comment (Andreagit97) [reply pkg/mutate/manage_nodeaffinity.go:19]: yes it is already merged, but I'll make it clearer in the next commit.

File: pkg/liqonet/natmappinginflater/natMappingInflater_test.go [hunk 5]
```diff
@@ -13,7 +13,7 @@ import (
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/dynamic/fake"
 
-	netv1alpha1 "github.com/liqotech/liqo/apis/net/v1alpha1"
+	liqonetapi "github.com/liqotech/liqo/apis/net/v1alpha1"
```
Comment (aleoli) [inline pkg/liqonet/natmappinginflater/natMappingInflater_test.go:16]: Why did you change the import name?

File: pkg/mutate/manage_pod_mutation.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+// createTolerationFromNamespaceOffloading creates a new virtualNodeToleration in case of LocalAndRemotePodOffloadingStrategyType
+// or RemotePodOffloadingStrategyType. In case of LocalPodOffloadingStrategyType returns an empty Toleration.
+func createTolerationFromNamespaceOffloading(strategy offv1alpha1.PodOffloadingStrategyType) corev1.Toleration {
+	var toleration corev1.Toleration
+	switch {
+	case strategy == offv1alpha1.LocalAndRemotePodOffloadingStrategyType, strategy == offv1alpha1.RemotePodOffloadingStrategyType:
+		// The virtual-node toleration must be added.
+		toleration = getVirtualNodeToleration()
+	default:
```
Comment (aleoli) [inline pkg/mutate/manage_pod_mutation.go:28]: In my opinion, we should add a case for `LocalPodOffloadingStrategyType` and print and error in the default case.

This should prevent future errors, for example, if we will add other strategies and we will forget to handle them in this switch.

File: pkg/mutate/manage_pod_mutation.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+		}
+	case nsoff.Spec.PodOffloadingStrategy == offv1alpha1.LocalAndRemotePodOffloadingStrategyType:
+		nodeSelector = nsoff.Spec.ClusterSelector
+		// A new NodeSelectorTerm terms that allows scheduling the pod also on local nodes is added.
+		newNodeSelectorTerm := corev1.NodeSelectorTerm{
+			MatchExpressions: append([]corev1.NodeSelectorRequirement{}, corev1.NodeSelectorRequirement{
+				Key:      liqoconst.TypeLabel,
+				Operator: corev1.NodeSelectorOpNotIn,
+				Values:   []string{liqoconst.TypeNode},
+			}),
```
Comment (aleoli) [inline pkg/mutate/manage_pod_mutation.go:60]:
```suggestion
			MatchExpressions: []corev1.NodeSelectorRequirement{
				{
					Key:      liqoconst.TypeLabel,
					Operator: corev1.NodeSelectorOpNotIn,
					Values:   []string{liqoconst.TypeNode},
				}
			}),
```

File: pkg/mutate/manage_pod_mutation.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+// createNodeSelectorFromNamespaceOffloading creates the right NodeSelector according to the PodOffloadingStrategy chosen.
+func createNodeSelectorFromNamespaceOffloading(nsoff *offv1alpha1.NamespaceOffloading) corev1.NodeSelector {
+	var nodeSelector corev1.NodeSelector
+	switch {
+	case nsoff.Spec.PodOffloadingStrategy == offv1alpha1.RemotePodOffloadingStrategyType:
+		// To ensure that the pod is not scheduled on local nodes is necessary to add a new NodeSelectorRequirement which
+		// requires the label "liqo.io/type=virtual-node", to every NodeSelectorTerm.
+		nodeSelector = nsoff.Spec.ClusterSelector
+		for i := range nodeSelector.NodeSelectorTerms {
+			nodeSelector.NodeSelectorTerms[i].MatchExpressions = append(nodeSelector.NodeSelectorTerms[i].MatchExpressions,
```
Comment (aleoli) [inline pkg/mutate/manage_pod_mutation.go:45]: Can you add comments on the reason why you are adding this expression to each term?

File: pkg/mutate/manage_pod_mutation.go [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+		nodeSelector.NodeSelectorTerms = append(nodeSelector.NodeSelectorTerms, newNodeSelectorTerm)
+	case nsoff.Spec.PodOffloadingStrategy == offv1alpha1.LocalPodOffloadingStrategyType:
+		// In local strategy there is no necessity to impose a particular NodeSelector.
+		nodeSelector = corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}}
+	}
+	return nodeSelector
+}
+
+// fillPodWithImposedNodeSelector gets the previously computed NodeSelector imposed by the PodOffloadingStrategy and
+// merges it with the Pod NodeSelector if it is already present. It simply adds it to the Pod if there is merge to do.
```
Comment (giorio94) [inline pkg/mutate/manage_pod_mutation.go:71]:
```suggestion
// merges it with the Pod NodeSelector if it is already present. It simply adds it to the Pod if previously unset.
```

File: pkg/mutate/manage_pod_mutation.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+func fillPodWithImposedNodeSelector(imposedNodeSelector *corev1.NodeSelector, pod *corev1.Pod) {
+	// To preserve the Pod Affinity content, it is necessary to add the imposedNodeSelector according to what
+	// is already present in the Pod Affinity.
+	switch {
+	case pod.Spec.Affinity == nil:
+		pod.Spec.Affinity = &corev1.Affinity{
+			NodeAffinity: &corev1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{},
```
Comment (giorio94) [inline pkg/mutate/manage_pod_mutation.go:79]: You could directly assign the `imposedNameSelector` here, instead of creating a placeholder. The same below

File: pkg/mutate/manage_pod_mutation.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+		oldNodeSelector := *pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution
+		tmpSelector := corev1.NodeSelector{NodeSelectorTerms: []corev1.NodeSelectorTerm{}}
+		for i := range oldNodeSelector.NodeSelectorTerms {
+			for j := range imposedNodeSelector.NodeSelectorTerms {
+				newMatchExpression := append(imposedNodeSelector.NodeSelectorTerms[j].MatchExpressions, oldNodeSelector.NodeSelectorTerms[i].MatchExpressions...)
+				tmpSelector.NodeSelectorTerms = append(tmpSelector.NodeSelectorTerms, corev1.NodeSelectorTerm{
+					MatchExpressions: newMatchExpression,
+				})
+			}
+		}
```
Comment (giorio94) [inline pkg/mutate/manage_pod_mutation.go:98]: IMHO, this logic deserves to be extracted to a separate function, for the sake of clarity and to simplify testing. More comments should also be added.

File: pkg/mutate/webhook_test.go [hunk 12]
```diff
@@ -0,0 +1,617 @@
...
+							Values:   []string{"C,D"},
+						}, {
+							Key:      "NotProvider",
+							Operator: corev1.NodeSelectorOpExists,
+						},
+						},
+					},
+				},
+				},
+			},
```
Comment (giorio94) [inline pkg/mutate/webhook_test.go:80]: It seems to me this is copyed and pasted three times, while changing the `PodOffloadingStrategy` only. You should use a function to generate the three items, avoiding code duplication and highlighting the differences.

File: pkg/mutate/webhook_test.go [hunk 13]
```diff
@@ -0,0 +1,617 @@
...
+						Key:      "NotProvider",
+						Operator: corev1.NodeSelectorOpExists,
+					}, {
+						Key:      liqoconst.TypeLabel,
+						Operator: corev1.NodeSelectorOpIn,
+						Values:   []string{liqoconst.TypeNode},
+					},
+					},
+				},
+			}},
```
Comment (giorio94) [inline pkg/mutate/webhook_test.go:184]: This could also be moved to a separate function (leveraged by the one above). Similar for the other occurrences of duplicated code below.

File: pkg/mutate/webhook_test.go [hunk 14]
```diff
@@ -0,0 +1,617 @@
...
+
+		It("Test that the PodAffinity in the case 2 is preserved", func() {
+			newPod := pods[2].DeepCopy()
+			newNodeSelector := imposedNodeSelector.DeepCopy()
+			fillPodWithImposedNodeSelector(newNodeSelector, newPod)
+			Expect(*newPod.Spec.Affinity.PodAffinity).To(Equal(podAffinity))
+		})
+	})
+
+	Context("4 - Call the mutatePod function and observe is pod is correctly mutated", func() {
```
Comment (giorio94) [inline pkg/mutate/webhook_test.go:449]:
```suggestion
	Context("4 - Call the mutatePod function and observe the pod is correctly mutated", func() {
```

File: pkg/liqonet/natmappinginflater/natMappingInflater_test.go (hunk 5 above)
Comment (Andreagit97) [reply pkg/liqonet/natmappinginflater/natMappingInflater_test.go:16]: GCI tool reported to me there were 2 identical imports called with a different name. So I have removed one of these. I don't know why in the old version it only reports one now. Maybe I will try to change again the name. Thank you

File: pkg/mutate/manage_pod_mutation.go (hunk 6 above)
Comment (Andreagit97) [reply pkg/mutate/manage_pod_mutation.go:28]: Great idea, thank you

File: pkg/mutate/manage_pod_mutation.go (hunk 8 above)
Comment (Andreagit97) [reply pkg/mutate/manage_pod_mutation.go:47]: I have modified a little bit the previous comment, I hope it is clearer now.

File: pkg/consts/namespace_mapping.go [hunk 15]
```diff
@@ -8,6 +8,7 @@ const (
 	// TypeLabel is the key of a Liqo label that identifies different types of nodes.
 	TypeLabel = "liqo.io/type"
 	// TypeNode is the value of a Liqo label that identifies Liqo virtual nodes.
+	// todo: change to VirtualNodeType
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/consts/namespace_mapping.go:11]: it will be addressed in the PR that will activate the new Namespace Mapping logic

File: cmd/liqo-webhook/main.go [hunk 16]
```diff
@@ -13,12 +13,16 @@ func main() {
 
 	setOptions(config)
 
-	log.Println("Starting server ...")
+	klog.Info("Starting server ...")
 
-	s, err := mutate.NewMutationServer(config)
+	ctx, cancel := context.WithCancel(context.Background())
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply cmd/liqo-webhook/main.go]: thank you I have adopted this approach, it should work

File: pkg/consts/webhook.go [hunk 17]
```diff
@@ -0,0 +1,7 @@
+package consts
+
+const (
+	// VirtualNodeTolerationKey all Pods that have to be scheduled on virtual nodes must have this toleration
+	// to Liqo taint.
+	VirtualNodeTolerationKey = "virtual-node.liqo.io/not-allowed"
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/consts/webhook.go:6]: I only found one recurrence of this constant throughout the whole project, the one that you have indicated above.

File: pkg/mutate/server.go [hunk 18]
```diff
@@ -18,13 +27,56 @@ type MutationServer struct {
 	mux    *http.ServeMux
 	server *http.Server
 
-	config *MutationConfig
+	webhookClient client.Client
+	config        *MutationConfig
 }
 
-func NewMutationServer(c *MutationConfig) (*MutationServer, error) {
+func getWebhookClient(ctx context.Context) client.Client {
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/mutate/server.go]: you are right

File: pkg/mutate/manage_pod_mutation.go [hunk 19]
```diff
@@ -0,0 +1,132 @@
...
+	klog.V(5).Infof("Generated Toleration: %s", toleration)
+
+	// Create the right NodeSelector according to the PodOffloadingStrategy case.
+	imposedNodeSelector := createNodeSelectorFromNamespaceOffloading(namespaceOffloading)
+	klog.V(5).Infof("ImposedNodeSelector: %s", imposedNodeSelector)
+
+	// If there is no toleration, return (LocalPodOffloadingStrategy case).
+	if toleration.MatchToleration(&corev1.Toleration{}) {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/mutate/manage_pod_mutation.go]: yes it is a good idea, thank you

File: pkg/mutate/manage_pod_mutation.go (hunk 11 above)
Comment (Andreagit97) [reply pkg/mutate/manage_pod_mutation.go]: you are right, I have created a new function

File: pkg/mutate/server.go [hunk 20]
```diff
@@ -18,13 +27,57 @@ type MutationServer struct {
 	mux    *http.ServeMux
 	server *http.Server
 
-	config *MutationConfig
+	webhookClient client.Client
+	config        *MutationConfig
+}
+
+func getWebhookClient(ctx context.Context) (client.Client, error) {
```
Comment (palexster) [inline pkg/mutate/server.go:34]: Can be this method factorized in the utils package externalizing the scheme? I think it can be reused outside the webhook.

**Review by palexster — APPROVED** [review]

**Review by giorio94 — COMMENTED** [review]
/lgtm

Just a couple of very minor nits inline.
I have a doubt concerning the server shutdown part, but it is better if we discuss in person about that.

File: pkg/mutate/manage_pod_mutation.go [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+				imposedNodeSelector)
+	}
+}
+
+// mutatePod checks the NamespaceOffloading CR associated with the Pod's Namespace.
+// The pod is modified in different ways according to the PodOffloadingStrategyType
+// chosen in the CR. Two possible modifications:
+// - The VirtualNodeToleration is added to the Pod Toleration if necessary.
+// - The old Pod NodeSelector is substituted with a new one according to the PodOffloadingStrategyType.
+func mutatePod(namespaceOffloading *offv1alpha1.NamespaceOffloading, pod *corev1.Pod) {
```
[the remark this answers is not in the document]
Comment (Andreagit97) [reply pkg/mutate/manage_pod_mutation.go:123]: good idea, thank you

File: pkg/mutate/server.go (hunk 20 above)
Comment (Andreagit97) [reply pkg/mutate/server.go:34]: Sure, I have created a new method in liqo utils.

File: pkg/utils/clientWithCache/get_client_with_cache.go [hunk 22]
```diff
@@ -0,0 +1,52 @@
+// Package clientwithcache contains utility methods to create a new controller runtime client with cache.
+package clientwithcache
```
Comment (giorio94) [inline pkg/utils/clientWithCache/get_client_with_cache.go:2]:
```suggestion
package cachedclient
```

Just a proposal though. The same for the filename.

**Review by giorio94 — APPROVED** [review]
/lgtm

Just squash the commits before merging
